Add mul() and umul() overloads with overflow detection

Adityazzzzz opened the pull request:
As requested by @rikkimax and @ibuclaw in Issue #22170.

This PR adds `mul` (signed) and `umul` (unsigned) overloads that accept an `out bool overflow` parameter.

**Changes:**
* Added `umul(Cent, Cent, out bool)` for unsigned overflow detection.
* Added `mul(Cent, Cent, out bool)` for signed overflow detection.
* Implemented using the division check method (`prod / a == b`) to verify results.
* Added unit tests covering signed and unsigned overflow edge cases (including `Cent.min` checks).

This allows callers to detect if a multiplication result has exceeded the 128-bit storage limit.

Adityazzzzz committed:
Add mul() and umul() overloads with overflow detection

ibuclaw commented:
> // Max Low * Max Low (2^64-1)^2 = 2^128 - 2^65 + 1. Should NOT overflow.
> res = umul(MAX_L, MAX_L, overflow);
> assert(!overflow);
> assert(res.hi == 0UL); // Should only populate low part (r0, r1)

@rikkimax again, a simple C __int128 program test shows that the bit representation is `{lo: 1, hi: ulong.max-1}`

ibuclaw commented:
> // 2. Signed Corner Cases (Magnitude and Zero/Identity)
> // S-Max Mag: 2^126 * 2 = 2^127. Valid magnitude for MinS. Should NOT overflow.
> res = smul(C2_126, C2, overflow);
> assert(!overflow);
> // Result should be exactly MinS (0x8000...)
> assert(res.lo == minS.lo && res.hi == minS.hi); 

@rikkimax and again.  `2^126` is a positive signed 128-bit number.  Signed `Cent.min` is a negative number.  So overflow _does_ occur.

At this point, I'd rather not go through the other examples, as they are up to this point all demonstrably false.

Adityazzzzz commented:
@WalterBright @ibuclaw @rikkimax  I have updated the implementation based on your feedback to avoid the expensive division check.
Changes in this update:

- Algorithm: Switched to a widening multiplication (calculating the full 128x128 -> 256-bit result) to detect overflow via the upper 128 bits ("excess"), as suggested by Walter.
- Implementation: Added a private umulExtended helper that decomposes operands into 64-bit parts to handle carries manually without using div.
- Signed Logic: Implemented the standard correction logic for signed multiplication (if (a < 0) excess -= b) to ensure the high bits are calculated correctly for signed inputs.
- Tests: Removed the previous generated tests and replaced them with explicit boundary checks (e.g., Cent.max, Cent.min, and cross-product overflows).

Please review

ibuclaw commented on druntime/src/core/int128.d:
```diff
@@ -502,6 +502,105 @@ Cent mul(ulong u1, ulong u2)
+    // Add p2 to middle
+    Cent p2_shifted = { lo: 0, hi: p2.lo }; // Declaration (OK)
+    res = add(res, p2_shifted);
+    if (ult(res, p2_shifted)) // Carry?
```
Possibly to help DMD, flip it around so there's one less function to inline.
```suggestion
    if (ugt(p2_shifted, res)) // Carry?
```

ibuclaw commented on druntime/src/core/int128.d:
```diff
@@ -1052,4 +1151,42 @@ unittest
+
+    // Small numbers -> No Overflow
+    res = umul(C10, C3, overflow);
+    assert(!overflow);
```
```suggestion
    assert(!overflow);
    assert(res == C30);
```

ibuclaw commented on druntime/src/core/int128.d:
```diff
@@ -502,6 +502,105 @@ Cent mul(ulong u1, ulong u2)
+    if (ult(res, p1_shifted)) // Carry?
+        excess = inc(excess);
+    
+    // FIX: Use Cent(...) constructor instead of { ... } inside the function call
```
```suggestion
```
Don't need the `FIX` comment.

Adityazzzzz committed:
fix: Use named arguments for Cent and add overflow assertion

WalterBright commented on druntime/src/core/int128.d:
```diff
@@ -502,6 +502,101 @@ Cent mul(ulong u1, ulong u2)
+ *      c1 * c2
+ */
+pure
+Cent umul(Cent c1, Cent c2, out bool overflow)
```
`overflow` should be `ref`, not `out`. The reason is so one can do this:
```
bool overflow = false;
i = mul(a,b,overflow);
k = mul(i,b,overflow);
j = mul(k,b,overflow);
if (overflow) printf("Danger, Will Robinson!\n");
```
I.e. one only has to check `overflow` once after cascaded calls, rather than after each operation.

WalterBright commented on druntime/src/core/int128.d:
```diff
@@ -502,6 +502,101 @@ Cent mul(ulong u1, ulong u2)
+ *      c1 * c2
+ */
+pure
+Cent mul(Cent c1, Cent c2, out bool overflow)
```
`ref` instead of `out`

WalterBright commented on druntime/src/core/int128.d:
```diff
@@ -502,6 +502,101 @@ Cent mul(ulong u1, ulong u2)
+    // If result is negative, excess must be -1 (all ones).
+    // If result is positive, excess must be 0 (all zeros).
+    if (cast(I)res.hi < 0)
+        overflow = tst(com(excess)); // ~excess != 0
```
|= instead of =, others as well

Adityazzzzz commented:
@ibuclaw yes, i missed the walter's comments, i will look into it

Adityazzzzz committed:
Fix review comments: use ref overflow and |= assignment

Adityazzzzz committed:
Merge branch 'master' into feature/mul-overflow
